frontend/readcollection.lua

```diff
@@ -2,149 +2,220 @@ local DataStorage = require("datastorage")
-    local dir
-    local should_write = false
-    if is_dir then
-        path = path .. "/"
-    end
-    local coll = self:readAllCollection()
-    for i in pairs(coll) do
-        local single_collection = coll[i]
-        for item = #single_collection, 1, -1 do
-            if not is_dir and single_collection[item].file == path then
-                should_write = true
-                table.remove(single_collection, item)
-            elseif is_dir then
-                dir = util.splitFilePathName(single_collection[item].file)
-                if dir == path then
-                    should_write = true
-                    table.remove(single_collection, item)
-                end
+end
+
+function ReadCollection:write(collection_name)
+    local collections = LuaSettings:open(collection_file)
+    for coll_name, coll in pairs(self.coll) do
+        if not collection_name or coll_name == collection_name then
+            local data = {}
+            for _, item in pairs(coll) do
+                table.insert(data, { file = item.file, order = item.order })
             end
+            collections:saveSetting(coll_name, data)
         end
     end
-    if should_write then
-        local collection = LuaSettings:open(collection_file)
-        collection.data = coll
-        collection:flush()
-    end
-end
-
-function ReadCollection:updateItemByPath(old_path, new_path)
-    local is_dir = false
-    local dir, file
-    if lfs.attributes(new_path, "mode") == "directory" then
-        is_dir = true
-        old_path = old_path .. "/"
-    end
-    local should_write = false
-    local coll = self:readAllCollection()
-    for i, j in pairs(coll) do
-        for k, v in pairs(j) do
-            if not is_dir and v.file == old_path then
-                should_write = true
-                coll[i][k].file = new_path
-            elseif is_dir then
-                dir, file = util.splitFilePathName(v.file)
-                if dir == old_path then
-                    should_write = true
-                    coll[i][k].file = string.format("%s/%s", new_path, file)
-                end
+    logger.dbg("ReadCollection: writing to collection file")
+    collections:flush()
+end
+
+function ReadCollection:getFileCollectionName(file, collection_name)
+    file = FFIUtil.realpath(file) or file
+    for coll_name, coll in pairs(self.coll) do
+        if not collection_name or coll_name == collection_name then
+            if coll[file] then
+                return coll_name, file
             end
         end
     end
-    if should_write then
-        local collection = LuaSettings:open(collection_file)
-        collection.data = coll
-        collection:flush()
-    end
 end
 
-function ReadCollection:removeItem(item, collection_name)
-    local coll = self:read(collection_name)
-    for k, v in pairs(coll) do
-        if v.file == item then
-            table.remove(coll, k)
-            break
+function ReadCollection:isAdded(file, collection_name)
```

Comment:
Reminds me of https://github.com/koreader/koreader/pull/11108#discussion_r1392562544 :)
`:hasFile()` ?